chamge/amqp/amqp-rpc.c

```diff
@@ -0,0 +1,700 @@
+/**
+ *  Copyright 2019 SK Telecom, Co., Ltd.
+ *    Author: Heekyoung Seo <hkseo@sk.com>
+ *
+ */
+
+#include "amqp-rpc.h"
+
+#include <amqp.h>
+#include <amqp_tcp_socket.h>
+
+#define DEFAULT_PORT 5672
+#define DEFAULT_CHANNEL 1
+#define DEFAULT_CONTENT_TYPE "application/json"
+
+typedef enum
+{
+  CHAMGE_AMQP_STATUS_NULL = 0,  /* initial state */
+  CHAMGE_AMQP_STATUS_READY,     /* socket created */
+  CHAMGE_AMQP_STATUS_CONNECTED, /* login finshed */
+  CHAMGE_AMQP_STATUS_SUBSCRIBING,       /* ready to read message */
+  CHAMGE_AMQP_STATUS_RECEIVED,  /* message received */
+} ChamgeAmqpRpcStatus;
+
+typedef struct amqp_connection_info rpc_connection_info;
+
+typedef struct _ChamgeAmqpRpcPrivate ChamgeAmqpRpcPrivate;
+struct _ChamgeAmqpRpcPrivate
+{
+  g_autofree gchar *amqp_url;
+
+  rpc_connection_info connection_info;
+  guint channel;
+
+  ChamgeAmqpRpcStatus state;
+
+  amqp_connection_state_t amqp_conn;
+  amqp_socket_t *amqp_socket;
+
+  GList *bindingkey_list;
+
+} ChamgeAmqpRcpPrivate;
+
+
+typedef enum
+{
+  PROP_HOSTNAME = 1,
+  PROP_PORT,
+  PROP_VHOST,
+  PROP_ID,
+  PROP_PWD,
+  PROP_CHANNEL,
+  PROP_URL,
+  PROP_LAST = PROP_URL
+} _ChamgeAmqpRpcProperty;
+
+static GParamSpec *properties[PROP_LAST + 1];
+
+/* *INDENT-OFF* */
+G_DEFINE_TYPE_WITH_CODE (ChamgeAmqpRpc, chamge_amqp_rpc, G_TYPE_OBJECT,
+    G_ADD_PRIVATE (ChamgeAmqpRpc))
+/* *INDENT-ON* */
+
+static ChamgeReturn
+_amqp_declare_queue (const amqp_connection_state_t conn,
+    const gchar * queue_name, amqp_bytes_t * queue)
+{
+  amqp_queue_declare_ok_t *amqp_r = NULL;
+  gchar *response_queue_name = NULL;
+
+  if (queue_name == NULL) {
+    amqp_r = amqp_queue_declare (conn, 1, amqp_empty_bytes, 0, 0, 0, 1,
+        amqp_empty_table);
+  } else {
+    amqp_r =
+        amqp_queue_declare (conn, 1, amqp_cstring_bytes (queue_name), 0, 0, 0,
+        1, amqp_empty_table);
+  }
+  if (amqp_r == NULL) {
+    g_error ("Can not declare a new queue or queue does not exist...");
+    return CHAMGE_RETURN_FAIL;
+  }
+  amqp_get_rpc_reply (conn);
+
+  g_assert_nonnull (amqp_r->queue.bytes);
+  response_queue_name = g_strdup (amqp_r->queue.bytes);
+  *queue = amqp_cstring_bytes (response_queue_name);
+
+  g_debug ("create private reply_to queue : name -> %s",
+      (gchar *) queue->bytes);
+  return CHAMGE_RETURN_OK;
+}
+
+static ChamgeReturn
+_amqp_property_set (amqp_basic_properties_t * const props,
+    const guint correlation_id, const gchar * reply_queue_name)
+{
+  gchar *correlation_id_str = NULL;
+  gchar *content_type_str = g_strdup (DEFAULT_CONTENT_TYPE);
+
+  props->_flags = AMQP_BASIC_CONTENT_TYPE_FLAG | AMQP_BASIC_DELIVERY_MODE_FLAG;
+  props->content_type = amqp_cstring_bytes (content_type_str);
+  props->delivery_mode = 2;     /* persistent delivery mode */
+
+  if (reply_queue_name != NULL) {
+    gchar *reply_to_queue = (gchar *) g_strdup (reply_queue_name);
```

Comment:
no casting is required for `g_strdup()`.

Reply:
Oops! ^^;